src/Voron/Impl/Journal/WriteAheadJournal.cs

```diff
@@ -2084,8 +2057,31 @@ private Pal.journal_entry PrepareToWriteToJournal(LowLevelTransaction tx, ref Pa
             var overhead = sizeOfPagesHeader + (long)numberOfPages * sizeof(long);
             var overheadInPages = checked((int)(overhead / Constants.Storage.PageSize + (overhead % Constants.Storage.PageSize == 0 ? 0 : 1)));
 
+            var freedPages = tx.GetFreedPages();
+
+            if (freedPages is { Count: > 0 })
+            {
+                // we need to remove the freed pages that were reused and modified later on during the transaction
+                
+                foreach (var page in txPages)
+                {
+                    if (freedPages.Remove(page.PageNumberInDataFile) && page.NumberOfPages > 1)
+                    {
+                        for (int i = 1; i < page.NumberOfPages; i++)
+                        {
+                            freedPages.Remove(page.PageNumberInDataFile + i);
+                        }
+                    }
```

Comment:
Shouldn't this be something that we are handling during the `ModifyPage`, instead?

Reply:
The problem is that it is entirely feasible to have a transaction in which we:

* modify the page
* free it
* allocate it again
* free it again

Etc...

That is a problem for this scenario, because you cannot assume that modified page / free pages are exclusive.